server/actions/serveFile.js

```diff
@@ -145,7 +151,44 @@ function serveStaticFile(req, res) {
     res.sendStatus(500);
   });
 
-  stream.pipe(res);
+  let chunks = [];
+  let codeString = "";
+  // IF HTML: push file content chunk-wise from stream to array
+  stream.on("data", chunk => {
+    if (html) {
+      chunks.push(chunk);
+    }
+  });
+
+  // IF HTML: return syntax highlighted code as html
+  stream.on("close", () => {
```

Comment:
An alternative (albeit a bit more complex) would be to make a transform stream to prepend and append the additional bits while still streaming the content. The main (only?) benefit here would be preventing you from reading the entire file into memory.

```js
const { Transform } = require('stream');

// sends content before and after another stream
function wrapStream(prefix, suffix) {
  let once = false;

  return new Transform({
    objectMode: true,
    transform(chunk, _, callback) {
      if (!once) {
        once = true;
        callback(null, `${prefix}${chunk}`);
      }
      else {
        callback(null, chunk);
      }
    },

    flush(callback) {
      callback(null, suffix);
    }
  });
}

// Actual usage

stream
  .pipe(wrapStream('<!doctype html><html><head>...<pre><code>', '</code></pre></body></html>'))
  .pipe(res)
;
```

Combined with my previous suggestion, this would change the code to look similar to:

```js
if (html) {
  stream
    .pipe(wrapStream('<!doctype html><html><head>...<pre><code>', '</code></pre></body></html>'))
    .pipe(res)
  ;
}
else {
  stream.pipe(res);
}
```

Reply:
I'll think about this :smile: